Add Arrow format doc (#948)
openvinotoolkit/datumaro: docs/source/docs/explanation/formats/arrow.md

# Arrow

## Format specification

> [Apache Arrow](https://arrow.apache.org/docs/format/Columnar.html) is a in-memroy columnar table format specification with multiple langauge support.

Reviewer: ```suggestion
[Apache Arrow](https://arrow.apache.org/docs/format/Columnar.html) is a in-memory columnar table format specification with multiple language support.
```

This format supports to export into a single arrow file or multiple shardings.

### Table schema

|   id   | subset | media.type | media.path | media.bytes | media.attributes | annotations | attributes |
|:------:|:------:|:----------:|:----------:|:-----------:|:----------------:|:-----------:|:----------:|
| string | string |   uint32   |   string   |    binary   |      binary      |    binary   |   binary   |

### id (`string`)
The ID of each entity. A tuple of (`id`, `subset`) is a unique key of each entity.

### subset (`string`)
The subset the entity belongs to. A tuple of (`id`, `subset`) is a unique key of each entity.

### media.type (`uint32`)
The type of media the entity has.

**Supported media types:**

- 0: `None`
- 2: `Image`
- 6: `PointCloud`

### media.path (`string`)
The path of the media. It could be a real path or a relative path, or `/NOT/A/REAL/PATH` if path is invalid.

### media.bytes (`binary`)
The binary data of the media. It could be `None` if one chooses not to save media when export.

### media.attributes (`binary`)
The attribute of the entity. The contents of it depends on `media.type`.
The byte order is little-endian.

**Image**

> ![image attributes](../../../../images/arrow/image_attributes.png)

Reviewer: Why don't we move this image directory to `docs/source/docs/explanation/formats`?

Author: I assumed that images are in top level as other images in ‘image’ directory. Do you want me to move them?

Reviewer: I think that's cleaner.

Author: Moved in https://github.com/openvinotoolkit/datumaro/pull/948/commits/187f2ad1c1f3b46346ba1939d8590911a1bd08d7

  ![bbox](../../../../images/arrow/shape.png)
- `Points`

  ![point](../../../../images/arrow/point.png)
- `Caption`

  ![caption](../../../../images/arrow/caption.png)
- `Cuboid3d`

  ![cuboid3d](../../../../images/arrow/cuboid3d.png)
- `Ellipse`

  ![ellipse](../../../../images/arrow/shape.png)

### attributes (`binary`)
The attributes of the entity. The byte order is little-endian.

![attributes](../../../../images/arrow/attributes.png)

## Import Arrow dataset

A Datumaro project with a Arrow source can be created in the following way:

```console
datum create
datum import --format arrow <path/to/dataset>
```

It is possible to specify project name and project directory. Run
`datum create --help` for more information.

An Arrow dataset directory should have the following structure:

<!--lint disable fenced-code-flag-->

```
└─ Dataset/
    ├── <subset_name_1>-0-of-2.arrow
    ├── <subset_name_1>-1-of-2.arrow
    ├── <subset_name_2>-0-of-1.arrow
    └── ...
```

If your dataset is not following the above directory structure,
it cannot detect and import your dataset as the Arrow format properly.

To make sure that the selected dataset has been added to the project, you can

> run `datum project info`, which will display the project information.

Reviewer: ```suggestion
run `datum project pinfo`, which will display the project information.
```

## Export to other formats

It can convert Datumaro dataset into any other format [Datumaro supports](/docs/data-formats/supported_formats/).
To get the expected result, convert the dataset to formats
that support the specified task (e.g. for panoptic segmentation - VOC, CamVID)

There are several ways to convert a Datumaro dataset to other dataset formats
using CLI:

- Export a dataset from Datumaro format to VOC format:

```console

> datum create
> datum import -f arrow <path/to/dataset>
> datum export -f voc -o <output/dir>

Reviewer: ```suggestion
datum project create
datum project import -f arrow <path/to/dataset>
datum project export -f voc -o <output/dir>
```

Author: I copied them from datumaro_binary.md. Does it also need to be changed? I will, if so.

Reviewer: We need to do it for all formats in a separate PR.

```

or

```console
datum convert -if arrow -i <path/to/dataset> -f voc -o <output/dir>
```

Or, using Python API:

```python
import datumaro as dm

dataset = dm.Dataset.import_from('<path/to/dataset>', 'arrow')
dataset.export('save_dir', 'voc', save_media=True, image_ext="AS-IS", num_workers=4)
```

## Export to Arrow

There are several ways to convert a dataset to Arrow format:

- Export a dataset from an existing project to Arrow format:
```console
# export dataset into Arrow format from existing project

> datum export -p <path/to/project> -f arrow -o <output/dir> \

Reviewer: ```suggestion
datum project export -p <path/to/project> -f arrow -o <output/dir> \
```
